ships/0029-one-off-builds.md

```diff
@@ -0,0 +1,228 @@
+for the sake of referencing them in a follow-up buildrun without the intention of creating multiple runs.
+
+A user might try to build a repository remotely in a `docker build` style and a single run is enough. This can be
+combined with the `local` source builds to give the user a single ergonomic command to push their source code and build the image in one step.
+
+The one-off build removes the burden of maintenance and cleanup of multiple resources for users and tools that do not
+profit from the re-usability of builds.
+
+### Goals
+
+- Execute builds without explicitly creating a `Build` resource
+- Extend `BuildRun` resource to allow embedding of the build specification
+- Extend `BuildRun` reconciler to pickup build specifications and translate them into `TaskRuns` directly
+- Extend shp cli to create a buildrun with an embedded build specification
+- Offer the same level of validation for embedded builds as regular builds
+
+### Non-Goals
+
+this ship focuses on `BuildRun` that embeds builds not vice versa
+
+- extend `Build` resource to start a "one-off" buildrun
+
+## Proposal
+
+We propose to extend the `BuildRun` API with a new field `buildSpec`. It is equivalent to the `spec` field in a `Build`
+resource and allows the same granularity of configuration. Thus, the enhancement is also referred to as an embedded build.
+The goal of this proposal are to avoid creating an actual build as a side-effect and to translating the build specification
+directly into a `TaskRun` without losing any of the validation steps for registering a build. This implies specifically
+that an embedded build spec does not correspond to any build resource in the cluster, nor are we proposing to create one
+in the process (more in alternatives). We also propose to not count embedded build specifications as builds in
+`shipwright/build` metrics.
+
+The following describes how such a `BuildRun` object can be specified:
+
+```yaml
+apiVersion: shipwright.io/v1alpha1
+kind: BuildRun
+metadata:
+  generateName: buildpack-nodejs-buildrun-
+spec:
+  buildSpec:
+      source:
+          url: https://github.com/shipwright-io/sample-nodejs
+          contextDir: source-build
+      strategy:
+          name: buildpacks-v3
+          kind: ClusterBuildStrategy
+      output:
+          image: docker.io/${REGISTRY_ORG}/sample-nodejs:latest
+          credentials:
+              name: push-secret
+```
+
+When an embedded specification is invalid, we propose not to start a `TaskRun`. Instead, the buildrun should be set
+to a failure state and the reasons for the invalid specification should be reflected using the `Succeded` condition in `status.conditions`.
+
+Additionally, we require that `buildSpec` is never used in conjunction with any of `buildRef`, `output`,
```

Comment:
To complete your list: paramValues, env, timeout.

I think it makes sense to generalize the statement like this: `buildSpec` is not meant to be used in conjunction `buildRef` and with any other field inside the BuildRun spec which is meant to override behavior of a Build. These are today: `env`, `output`, `paramValues`, `sources`, and `timeout`. Other fields such as `serviceAccount` are allowed in conjunction with `buildSpec`.